### PR TITLE
File browser menu

### PR DESCRIPTION
**Pull request by yatli**
POC, not working yet

### CODE REVIEW

**Comment by yatli** [comment]
finally figured out how to specify extra build flags in my environment. (sorry!)

**Comment by yatli** [comment]
after these changes, the compiler starts reporting undefined references to LightPage::display()

**Comment by jmamma** [comment]
when linking or compiling?

**Comment by yatli** [comment]
linking.

Indeed LightPage::display is declared but not defined, but I don't understand why it starts reporting the error, despite the changes are not directly related to LightPage.

**Comment by jmamma** [comment]
This sounds familiar...

**Comment by yatli** [comment]
Last night I then went ahead to define an empty display() {}, but the result firmware didn't boot up, stuck at "SD Initializing".

I thought I finally trashed the arduino and swapped a new one but the problem persisted, so I decided to get some sleep first :)

**Comment by jmamma** [comment]
deb30a9  Does that fix it ?

**Comment by yatli** [comment]
it compiles and links.
still not booting up, but I doubt it's a separate problem.

**Comment by yatli** [comment]
I'm going over all the virtual functions to try to catch all the undefined ones.

**Comment by jmamma** [comment]
I will check in a minute. Had linking problems of my own here.

**Comment by yatli** [comment]
PageParent::update is not defined.

**Comment by yatli** [comment]
LightPage::lockEncoders

**Comment by yatli** [comment]
LightPage::unlockEncoders

File: avr/cores/megacommand/GUI/Pages.hh [hunk 1]
```diff
@@ -182,13 +182,8 @@ public:
   virtual void update();
   /** This will clear the encoder movements. **/
   virtual void clear();
-  /** Display the page using the display routine as specfied by function
-   * pointer*/
-  virtual void display();
```
Comment (jmamma) [inline avr/cores/megacommand/GUI/Pages.hh:187]: Surely it has to be these changes that are causing the issue.

**Comment by yatli** [comment]
ok, that's all of them.

The AutoEncoder thing is suspicious, looking into it

**Comment by yatli** [comment]
not booting up after https://github.com/jmamma/MIDICtrl20_MegaCommand/pull/49/commits/b9fc37c61c14c5634eff0f0cc2c70e850762c6fb

trying a minimal repro

**Comment by yatli** [comment]
base setup: all undefined method changes, no menu POC code = OK

**Comment by yatli** [comment]
adding changes in `Menu.h`, `Menu.cpp` = OK

**Comment by yatli** [comment]
adding `FileBrowserPage.cpp`, `FileBrowserPage.h` = NOT BOOTING
the problem is isolated.

**Comment by jmamma** [comment]
I'm compiling now.

**Comment by jmamma** [comment]
Compiles and boots for me.

The Grid page is not being displayed correctly. No text

**Comment by jmamma** [comment]
Not sure what your compiling environment is like.
But I run a  ./clean.sh before compiling to remove any stale object files. 
Note: Changing header will not cause recompile of underlying .c  

--

Whenever I have link or weird behaviour it's usually because old object files have been linked in.

**Comment by jmamma** [comment]
```git checkout dev Pages.hh```

== link problem.

I'm wondering if this is it:

```
+  Encoder* param1;
+  Encoder* param2;
 
   FileBrowserPage(Encoder *e1 = NULL, Encoder *e2 = NULL, Encoder *e3 = NULL,
                   Encoder *e4 = NULL)
-      : LightPage(e1, e2, e3, e4) {}
+      : LightPage(e1, e2, e3, e4), param1(e1), param2(e2) {}

```
That looks strange to me.

**Comment by jmamma** [comment]
```
09 class Encoder : public EncoderParent {
110  /** Short name. **/
111  public:
112 
113     
114     Encoder(const char *_name = NULL, encoder_handle_t _handler = NULL);
```

**Comment by yatli** [comment]
`param1(e1)` is a shorthand for writing `param1 = e1;` in the constructor

**Comment by yatli** [comment]
I need to keep track of the original encoder setup by derived class because if I install the file menu encoders, the originals would be lost

But nevertheless, I'll try remove these and see what happens.

**Comment by jmamma** [comment]
hmm. 😭

**Comment by yatli** [comment]
> Changing header will not cause recompile of underlying .c

Add a dependency from .cpp to .h solves the problem.

**Comment by jmamma** [comment]
I've reverted all the FileBrowser pages code to dev. and still getting the link problem.

**Comment by jmamma** [comment]
Found him.

**Comment by jmamma** [comment]
```
+++ b/avr/cores/megacommand/MCL/Menu.h
@@ -37,7 +37,9 @@ public:
   uint8_t values[MAX_MENU_ITEMS];
   uint8_t entry_mask[(MAX_MENU_ITEMS + 7) / 8];
 
+  Menu(){ 
+          memset(entry_mask, 0xFF, sizeof(entry_mask));
+        }
 
```

**Comment by jmamma** [comment]
@yatli any idea why that line is causing 💔

**Comment by jmamma** [comment]
void * memset ( void * ptr, int value, size_t num );

arg1 should be a pointer. ✅

**Comment by yatli** [comment]
@jmamma I'm still trying.
If I revert `Menu.h`, `Menu.cpp` it boots

**Comment by jmamma** [comment]
I better take a break. I'm making no positive contribution to this debugging effort.

**Comment by yatli** [comment]
don't worry, I'll figure it out (eventually) :)

**Comment by yatli** [comment]
@jmamma I believe I've found the root cause. It's still not working but I think it's on the right path.

**Comment by yatli** [comment]
Currently the code is at some kind of capacity limit boundary. I have the new menu code, most boilerplates in FileBrowserPage, and the encoders.

Up to this point, everything works as expected.

Then I continue to uncomment `MenuPage file_menu_page(&file_menu_layout, &config_param1, &file_menu_encoder);`, and things are getting weird, including corrupted font. The corrupted bitmaps of the font look like some kind of data so perhaps it's overlapping with other content.

**Comment by yatli** [comment]
And finally, after the (trivial looking) last commit above, the firmware becomes unbootable.
It doesn't look like any bad implementation, but the code size as gone pass some point.

**Comment by jmamma** [comment]
> Currently the code is at some kind of capacity limit boundary. I have the new menu code, most boilerplates in FileBrowserPage, and the encoders.
> 
> Up to this point, everything works as expected.
> 
> Then I continue to uncomment `MenuPage file_menu_page(&file_menu_layout, &config_param1, &file_menu_encoder);`, and things are getting weird, including corrupted font. The corrupted bitmaps of the font look like some kind of data so perhaps it's overlapping with other content.

You might be on to something. 

Can you disable WAV_DESIGNER from MCL.h and see if you have corruption with your code added.

**Comment by yatli** [comment]
I have disabled some other code in FileBrowserPage and it starts booting again!

**Comment by jmamma** [comment]
Out of curiosity. Are you using an official arduino board?

**Comment by yatli** [comment]
nope, some cheap Chinese replica.
but the MCU should be identical.

**Comment by jmamma** [comment]
I dont have my official board handy. Anyway, we can verify program upload use avrdude/ arduino ide. It should perform a checksum.

**Comment by yatli** [comment]
I don't think the board has reached its maximum code capacity -- a critically bootable source tree with DEBUGMODE on, is still bootable.

DEBUGMODE generally means +6% program storage space currently

**Comment by yatli** [comment]
Something must be wrong with PROGMEM

**Comment by yatli** [comment]
@jmamma https://forum.arduino.cc/index.php?topic=387506.0 does that ring the bell?

**Comment by jmamma** [comment]
digging through the forums now...

**Comment by jmamma** [comment]
https://www.avrfreaks.net/comment/600237#comment-600237

progmem is limited to 16bit address space.

**Comment by jmamma** [comment]
https://gist.github.com/Koepel/2a4ae36ae6397cd260f2c7ae2a50610e

**Comment by jmamma** [comment]
../../../../tools/avr/bin/avr-objdump -h main.hex

^ Have a play with object dump "-s" flag will dump hex contents. I've noticed there is a  lot of padded space at the start of the binary.

**Comment by yatli** [comment]
grep for `PROGMEM =` to get a list of those definitions.
There's a lot of unused font in the Adafruit library. We can probably trim them.

**Comment by jmamma** [comment]
Those fonts shouldn't be stored unless included.

**Comment by jmamma** [comment]
Take a look at the compiled .hex binary.
So many bytes wasted in the first 1/4 of the file where PROGMEM should be.

```
 1c70 00000000 00000000 00000000 00000000  ................
 1c80 00000000 00000000 00000000 00000000  ................
 1c90 00000000 00000000 00000000 00000000  ................
 1ca0 00000000 00000000 00000000 00000000  ................
 1cb0 00000000 00000000 00000000 00000000  ................
 1cc0 00000000 00000000 00000000 00000000  ................
 1cd0 00000000 00000000 00000000 00000000  ................
 1ce0 00000000 00000000 00000000 00000000  ................
 1cf0 00000000 00000000 00000000 00000000  ................
 1d00 00000000 00000000 00000000 00000000  ................
 1d10 00000000 00000000 00000000 00000000  ................
 1d20 00000000 00000000 00000000 00000000  ................
 1d30 00000000 00000000 00000000 00000000  ................
 1d40 00000000 00000000 00000000 00000000  ................
 1d50 00000000 00000000 00000000 00000000  ................
 1d60 00000000 00000000 00000000 00000000  ................
 1d70 00000000 00000000 00000000 00000000  ................
 1d80 00000000 00000000 00000000 00000000  ................
 1d90 00000000 00000000 00000000 00000000  ................
 1da0 00000000 00005359 5354454d 00000000  ......SYSTEM....
 1db0 00024449 53504c41 593a0000 00000000  ..DISPLAY:......
 1dc0 00000000 02026859 00000000 00494e54  ......hY.....INT
 1dd0 00000000 00000000 00000000 00000149  ...............I
 1de0 4e542b45 58540000 00000000 00000000  NT+EXT..........
 1df0 00000000 00000000 00000000 00000000  ................
 1e00 00000000 00000000 00000000 00000000  ................
 1e10 00000000 00000000 00000000 00000000  ................
```

**Comment by yatli** [comment]
```
 ┌─┐ C:\Users\Yatao\Documents\Arduino\hardware\MIDICtrl20_MegaCommand\sketch                                              menu_mask ◇ ┆ Yatao
 └─╜ ╰▷ & "C:\Program Files (x86)\Arduino\hardware\tools\avr\bin\avr-objdump.exe" -h .\sketch.MIDICtrl20_MegaCommand.avr.mega.hex

.\sketch.MIDICtrl20_MegaCommand.avr.mega.hex:     file format ihex

Sections:
Idx Name          Size      VMA       LMA       File off  Algn
  0 .sec1         00010000  00000000  00000000  00000000  2**0
                  CONTENTS, ALLOC, LOAD
  1 .sec2         00010000  00010000  00010000  0002d011  2**0
                  CONTENTS, ALLOC, LOAD
  2 .sec3         00010000  00020000  00020000  0005a022  2**0
                  CONTENTS, ALLOC, LOAD
  3 .sec4         00008596  00030000  00030000  00087033  2**0
                  CONTENTS, ALLOC, LOAD
```

The hex file consists of 4 sections (64KB max each)

**Comment by jmamma** [comment]
More info:

../../../../tools/avr/bin/avr-readelf -a main.elf

**Comment by yatli** [comment]
```
   166: 00000416  5040 OBJECT  LOCAL  DEFAULT    2 _ZL16file_menu_layout
   162: 000017c6  5040 OBJECT  LOCAL  DEFAULT    2 _ZL20rampage1_menu_layout
   160: 00002b76  5040 OBJECT  LOCAL  DEFAULT    2 _ZL21mclconfig_menu_layou
   158: 00003f26  5040 OBJECT  LOCAL  DEFAULT    2 _ZL17chain_menu_layout
   156: 000052d6  5040 OBJECT  LOCAL  DEFAULT    2 _ZL20mdconfig_menu_layout
   154: 00006686  5040 OBJECT  LOCAL  DEFAULT    2 _ZL22midiconfig_menu_layo
   152: 00007a36  5040 OBJECT  LOCAL  DEFAULT    2 _ZL18system_menu_layout
   151: 00008de6  5040 OBJECT  LOCAL  DEFAULT    2 _ZL21auxconfig_menu_layou
   176: 0000a196  5040 OBJECT  LOCAL  DEFAULT    2 _ZL16slot_menu_layout
   210: 0000b546  5040 OBJECT  LOCAL  DEFAULT    2 _ZL17track_menu_layout
```

how are they taking 5040 bytes each?

**Comment by yatli** [comment]
```
   386: 0000dc2a   665 OBJECT  LOCAL  DEFAULT    2 _ZL14TomThumbGlyphs
   387: 0000dec3   418 OBJECT  LOCAL  DEFAULT    2 _ZL15TomThumbBitmaps
   379: 0000e065     7 OBJECT  LOCAL  DEFAULT    2 _ZL11Elektrothic
   378: 0000e06c     7 OBJECT  LOCAL  DEFAULT    2 _ZL8TomThumb
   415: 0000e073   665 OBJECT  LOCAL  DEFAULT    2 _ZL14TomThumbGlyphs
   416: 0000e30c   418 OBJECT  LOCAL  DEFAULT    2 _ZL15TomThumbBitmaps
   414: 0000e4ae     7 OBJECT  LOCAL  DEFAULT    2 _ZL8TomThumb
   428: 0000e4b5   665 OBJECT  LOCAL  DEFAULT    2 _ZL14TomThumbGlyphs
   429: 0000e74e   418 OBJECT  LOCAL  DEFAULT    2 _ZL15TomThumbBitmaps
   425: 0000e8f0     7 OBJECT  LOCAL  DEFAULT    2 _ZL8TomThumb
   448: 0000e8f7  1280 OBJECT  LOCAL  DEFAULT    2 _ZL4font
  1603: 0000edf7   665 OBJECT  GLOBAL HIDDEN     2 _ZL14TomThumbGlyphs.lto_p
  1442: 0000f090   418 OBJECT  GLOBAL HIDDEN     2 _ZL15TomThumbBitmaps.lto_
```

tomthumb goes in 4 times...

edit: oh wait there's more of them below

**Comment by jmamma** [comment]
😱

**Comment by jmamma** [comment]
Time for sleep. 😨

**Comment by yatli** [comment]
Good night!

**Comment by jmamma** [comment]
> how are they taking 5040 bytes each?

function pointers in progmem might have something to do with it. there are some compile warnings as well.

**Comment by yatli** [comment]
@jmamma with the tomthumb dups removed we regain 5K pgmspace.

**Comment by yatli** [comment]
I have modified MenuPage, Menu and menu_t to leverage templates for dynamic sizes.
Program space is brought down from 82% to 66%. 😱

**Comment by yatli** [comment]
it's usable now. still need a few finishing touches.

**Comment by jmamma** [comment]
❤️ ❤️ ❤️ ❤️

**Comment by jmamma** [comment]
@yatli thank you for your enduring work here. 

I recompiled the binary and noticed we still have padding. The cause being:

```
  9 struct menu_option_t {
 10   uint8_t pos;
 11   char name[12];
 12 };
 13 
 14 struct menu_item_t {
 15   char name[14];
 16   uint8_t min;
 17   uint8_t range;
 18   uint8_t number_of_options;
 19   uint8_t *destination_var; // pointer to variable to be updated by param change
 20   Page *page_callback;
 21   void (*row_function)();
 22   menu_option_t options[MAX_MENU_OPTIONS];
```
MAX_MENU_OPTIONS = 16
sizeof(struct menu_option_t) = 13.
16 * 13 = 208 bytes per menu entry.

**Comment by jmamma** [comment]
I thought about ways of improving this further. We start running in to the problem of sacrificing code simplicity for space efficiency.

A simple optimisation here would be set MAX_MENU_OPTIONS = 6.
char name[12] to char name[7]

**Comment by jmamma** [comment]
../../../../tools/avr/bin/avr-objdump -S main.elf 

^ prints out the PROGMEM data structures + assembly code per function. Pretty cool for optimisation analysis.

**Comment by yatli** [comment]
`MAX_MENU_OPTIONS = 6` -- 2% space gain

**Comment by yatli** [comment]
A minor problem with the rename function -- the user must keep the suffix or the renamed entry will disappear because it doesn't match 0_0

**Comment by yatli** [comment]
Ok it's fully working now. Navigation is also improved.
Ready to merge now.

**Comment by jmamma** [comment]
Nice Work @yatli .

I can finally delete all my 100s of test projects.

**Comment by jmamma** [comment]
Just played with the SoundBrowser page.

Looking good!

This functionality is long overdue. Thanks so much.

**Comment by jmamma** [comment]
I wonder if we should consolidate New Project and Load Project menu items in a future patch.

Going to merge now.

**Pull request merged**